website/src/components/Chat/PluginsChooser.tsx

```diff
@@ -0,0 +1,257 @@
+  const { control, register, reset, setValue } = useFormContext<ChatConfigFormData>();
+  const [selectedPluginIndex, setSelectedPluginIndex] = useState<number | null>(null);
+  const { isOpen, onOpen, onClose } = useDisclosure();
+
+  const textareaRef = useRef<HTMLTextAreaElement | null>(null);
+  const containerRef = useRef<HTMLDivElement | null>(null);
+
+  const findSelectedPluginIndex = useCallback(() => {
+    return plugins.findIndex((plugin) => plugin.enabled);
+  }, [plugins]);
+
+  const handlePluginSelect = useCallback(
+    (index: number) => {
+      const selectedIndex = plugins.findIndex((plugin) => plugin.enabled);
+      const newPlugins = plugins.map((plugin, i) => ({
+        ...plugin,
+        enabled: selectedIndex === index ? false : i === index,
+      }));
+      setValue("plugins", newPlugins, { shouldDirty: true });
+      // onPluginActivated(newPlugins[index].enabled);
+    },
+    [setValue, plugins]
+  );
+
+  const handlePluginEdit = useCallback(
+    (index: number) => {
+      setSelectedPluginIndex(index);
+      onOpen();
+    },
+    [onOpen]
+  );
+
+  const handlePluginRemove = useCallback(
+    (index) => {
+      const updatedPlugins = plugins.filter((_, idx) => idx !== index);
+      setValue("plugins", updatedPlugins, { shouldDirty: true });
+    },
+    [setValue, plugins]
+  );
+
+  const handlePluginSave = useCallback(async () => {
+    if (textareaRef.current) {
+      const newPlugins = [...plugins];
+      newPlugins[selectedPluginIndex!].url = textareaRef.current.value;
+
+      const plugin: PluginEntry = await post(API_ROUTES.GET_PLUGIN_CONFIG, {
+        arg: { plugin: newPlugins[selectedPluginIndex!] },
+      });
+      // disable all existing plugins and enable new one
+      newPlugins.forEach((plugin) => (plugin.enabled = false));
+      newPlugins[selectedPluginIndex!].enabled = true;
+      newPlugins[selectedPluginIndex!].plugin_config = plugin.plugin_config;
+      setValue("plugins", newPlugins, { shouldDirty: true });
+      //onPluginActivated(newPlugins[selectedPluginIndex!].enabled);
+    }
+    onClose();
+  }, [onClose, control, plugins, selectedPluginIndex]);
+
+  const handlePluginAdd = useCallback(() => {
+    const newPlugin: PluginEntry = {
+      url: "",
+      enabled: false,
+    };
+    setValue("plugins", [...plugins, newPlugin]);
+    handlePluginEdit(plugins.length);
+  }, [setValue, plugins, handlePluginEdit]);
+
+  const limitPluginNameLength = (name: string) => {
+    if (!name) {
+      return "";
+    }
+    if (name.length > 15) {
+      return name.substring(0, 15) + "...";
+    }
+    return name;
+  };
+
+  return (
+    <FormControl mb="40px" ref={containerRef}>
+      <Box position="fixed" zIndex="1">
+        <Menu placement="auto">
+          <MenuButton as={Button} pr="3" rightIcon={<Cog />} w={measureDivWidth(containerRef.current) + 15} size="lg">
+            {plugins && plugins[findSelectedPluginIndex()]?.enabled ? (
```

Comment:
move `findSelectedPluginIndex` to the body of the function, maybe also `useMemo`

Reply:
done